### PR TITLE
EfficientDet text graph generator

### PR DESCRIPTION
**Pull request by dkurt**
Model from https://github.com/google/automl/

| DNN_BACKEND_OPENCV | DNN_BACKEND_INFERENCE_ENGINE |
|---|---|
|241.67ms | 88.53ms |

CPU: Intel(R) Core(TM) i5-4460  CPU @ 3.20GHz

**Merge with extra**: https://github.com/opencv/opencv_extra/pull/763

#### Prepare frozen graph
1. Download repository and checkout from https://github.com/google/automl/tree/master/efficientdet#2-pretrained-efficientdet-checkpoints
2. Run
```
python3 model_inspect.py --runmode=saved_model --model_name=efficientdet-d0  --ckpt_path=efficientdet-d0 --saved_model_dir=savedmodeldir
```
3. Optimize
```python
import tensorflow as tf
from tensorflow.python.tools import optimize_for_inference_lib
from tensorflow.tools.graph_transforms import TransformGraph

pb_file = 'savedmodeldir/efficientdet-d0_frozen.pb'
graph_def = tf.compat.v1.GraphDef()

try:
    with tf.io.gfile.GFile(pb_file, 'rb') as f:
        graph_def.ParseFromString(f.read())
except:
    with tf.gfile.FastGFile(pb_file, 'rb') as f:
        graph_def.ParseFromString(f.read())

graph_def = optimize_for_inference_lib.optimize_for_inference(graph_def, ['image_arrays'], ['detections'], tf.uint8.as_datatype_enum)
graph_def = TransformGraph(graph_def, ['image_arrays'], ['detections'], ['fold_constants'])

with tf.gfile.FastGFile('efficientdet-d0_opt.pb', 'wb') as f:
   f.write(graph_def.SerializeToString())
```

4. Generate a `.pbtxt` text graph by
```
python3 tf_text_graph_efficientdet.py --input efficientdet-d0_opt.pb --output efficientdet-d0_opt.pbtxt
```

5. Test
```python
import cv2 as cv

net = cv.dnn_DetectionModel('efficientdet-d0_opt.pb', 'efficientdet-d0_opt.pbtxt')
net.setInputSize(512, 512)
net.setInputScale(1.0 / 255)
net.setInputMean((123.675, 116.28, 103.53))

frame = cv.imread('example.jpg')

classes, confidences, boxes = net.detect(frame, confThreshold=0.5, nmsThreshold=0.4)
for classId, confidence, box in zip(classes.flatten(), confidences.flatten(), boxes):
    cv.rectangle(frame, box, color=(255, 0, 0), thickness=3)
    cv.rectangle(frame, box, color=(0, 255, 0), thickness=1)

cv.imshow('out', frame)
cv.waitKey()
```

```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2020.2.0:16.04
build_image:Custom Win=openvino-2020.2.0
build_image:Custom Mac=openvino-2020.2.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: modules/dnn/test/test_tf_importer.cpp [hunk 1]
```diff
@@ -1123,4 +1123,37 @@ TEST_P(Test_TensorFlow_nets, Mask_RCNN)
...
 }
 
+TEST_P(Test_TensorFlow_nets, EfficientDet)
+{
+    if (target != DNN_TARGET_CPU)
+    {
+        if (target == DNN_TARGET_OPENCL_FP16) applyTestTag(CV_TEST_TAG_DNN_SKIP_IE_OPENCL_FP16);
+        if (target == DNN_TARGET_OPENCL)      applyTestTag(CV_TEST_TAG_DNN_SKIP_IE_OPENCL);
+        if (target == DNN_TARGET_MYRIAD)      applyTestTag(CV_TEST_TAG_DNN_SKIP_IE_MYRIAD);
+    }
```
Comment (l-bat) [inline modules/dnn/test/test_tf_importer.cpp:1133]: Is test skipped for `OCV` and` DLIE` backends or only for `DLIE`?

Comment (dkurt) [reply modules/dnn/test/test_tf_importer.cpp:1133]: For OpenCV as well

Comment (l-bat) [reply modules/dnn/test/test_tf_importer.cpp:1133]: I think it is better to use  `CV_TEST_TAG_DNN_SKIP_OPENCL` and `CV_TEST_TAG_DNN_SKIP_OPENCL_FP16`

File: samples/dnn/tf_text_graph_efficientdet.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        if re.match('box_net/box-predict(_\d)*/separable_conv2d$', node.name):
+            node.addAttr('loc_pred_transposed', True)
+
+        # Replace RealDiv to Mul with inversed scale for compatibility
+        if node.op == 'RealDiv':
+            for inpNode in graph_def.node:
+                if inpNode.name != node.input[1] or not 'value' in inpNode.attr:
+                    continue
+
+                # print(inpNode.name, inpNode.attr['value']['tensor'])
```
Comment (l-bat) [inline samples/dnn/tf_text_graph_efficientdet.py:97]: Is this line needed?

File: samples/dnn/tf_text_graph_efficientdet.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+    parser.add_argument('--aspect_ratios', default=[1.0, 1.0, 1.4, 0.7, 0.7, 1.4],
+                        nargs='+', type=float, help='Parameter from training config')
+    parser.add_argument('--num_classes', default=90, type=int, help='Number of classes to detect')
+    parser.add_argument('--width', default=512, type=int, help='Network input width')
+    parser.add_argument('--height', default=512, type=int, help='Network input height')
+    args = parser.parse_args()
+
+    ar = args.aspect_ratios
+    assert(len(ar) % 2 == 0)
+    ar = [(ar[i*2], ar[i*2+1]) for i in range(len(ar)//2)]
```
Comment (l-bat) [inline samples/dnn/tf_text_graph_efficientdet.py:234]: `ar = zip(ar[::2], ar[1::2])` is better for me

Comment (dkurt) [reply samples/dnn/tf_text_graph_efficientdet.py]: Replaced. Thanks!

**Comment by dkurt** [comment]
Please do not merge - will make one mo change.

**Comment by rchglev** [comment]
Hello everyone, I am testing convert code with efficientdet-d0_frozen.pb.
I tested with TF 2.2.0
But faced this error
ModuleNotFoundError: No module named 'tensorflow.tools.graph_transforms'

Please tell me which tf version should I use.

I tested with tf 1.15.0 too
Also faced can not apply transformation of fold_constant.
How can I solve this issues.
Thank you.

**Comment by rchglev** [comment]
Traceback (most recent call last):
  File "D:/Work/EffectnetSample/automl-master/efficientdet/convertpb.py", line 16, in <module>
    graph_def = TransformGraph(graph_def, ['image_arrays'], ['detections'], ['fold_constants'])
 
I got this error when used TF1.6.0

 File "C:\python36\lib\site-packages\tensorflow\tools\graph_transforms\__init__.py", line 51, in TransformGraph
    transforms_string, status)
  File "C:\python36\lib\site-packages\tensorflow\python\framework\errors_impl.py", line 516, in __exit__
    c_api.TF_GetCode(self.status.status))
tensorflow.python.framework.errors_impl.InvalidArgumentError: NodeDef mentions attr 'batch_dims' not in Op<name=GatherV2; signature=params:Tparams, indices:Tindices, axis:Taxis -> output:Tparams; attr=Tparams:type; attr=Tindices:type,allowed=[DT_INT32, DT_INT64]; attr=Taxis:type,allowed=[DT_INT32, DT_INT64]>; NodeDef: GatherV2 = GatherV2[Taxis=DT_INT32, Tindices=DT_INT32, Tparams=DT_FLOAT, batch_dims=0](Const_5, strided_slice_9, GatherV2/axis). (Check whether your GraphDef-interpreting binary is up to date with your GraphDef-generating binary.).

**Comment by dkurt** [comment]
@rchglev, please try TensorFlow 1.15.0 once again. It should work

**Comment by rchglev** [comment]
Python 3.5?
or 3.6?
Now I am trying with 3.6.

**Comment by dkurt** [comment]
@rchglev it does not matter which python to use.

IF you need only D0, you may download already frozen graph by the link from https://github.com/opencv/opencv_extra/pull/763

**Comment by rchglev** [comment]
Thank you but I need environment of convert from frozen to optimize.
And also this code do not work too:
net = cv.dnn_DetectionModel('efficientdet-d0_opt.pb', 'efficientdet-d0_opt.pbtxt')
It tells me there is no this function.
I built and installed  your opencv branch already.
Please solve this issue
Best regards.
Capture

**Comment by alalek** [comment]
`cv.dnn_DetectionModel` is covered by Python tests, so it exists and works.
`cv.getBuildInformation()` can show which version you really get in your environment. Anyway we can't fix that.
*Hint*: `data` directory means that you have mess with unofficial `opencv-python` package from pip.

Build / usage questions should go to Users OpenCV Q/A forum: http://answers.opencv.org
This tracker is for issues and bugs that needs fix in OpenCV.

**Comment by rchglev** [comment]
Thank you alalek.

I wish you teach me the reason of this error:

Traceback (most recent call last):
File "D:/Work/EffectnetSample/automl-master/efficientdet/convertpb.py", line 16, in
graph_def = TransformGraph(graph_def, ['image_arrays'], ['detections'], ['fold_constants'])

I got this error when used TF1.6.0

File "C:\python36\lib\site-packages\tensorflow\tools\graph_transforms_init_.py", line 51, in TransformGraph
transforms_string, status)
File "C:\python36\lib\site-packages\tensorflow\python\framework\errors_impl.py", line 516, in exit
c_api.TF_GetCode(self.status.status))
tensorflow.python.framework.errors_impl.InvalidArgumentError: NodeDef mentions attr 'batch_dims' not in Op<name=GatherV2; signature=params:Tparams, indices:Tindices, axis:Taxis -> output:Tparams; attr=Tparams:type; attr=Tindices:type,allowed=[DT_INT32, DT_INT64]; attr=Taxis:type,allowed=[DT_INT32, DT_INT64]>; NodeDef: GatherV2 = GatherV2[Taxis=DT_INT32, Tindices=DT_INT32, Tparams=DT_FLOAT, batch_dims=0](Const_5, strided_slice_9, GatherV2/axis). (Check whether your GraphDef-interpreting binary is up to date with your GraphDef-generating binary.).
Thank you

**Comment by dkurt** [comment]
@rchglev, 
> I got this error when used TF1.6.0

Please use TensorFlow 1.15.0

**Comment by rchglev** [comment]
Python 3.6?
Thank you

**Pull request merged**